src/Face.cu

```diff
@@ -11,19 +10,49 @@ __device__ SpacePoint calculate_normal(const SpacePoint *vertices)
     return normal / get_distance(normal, origin);
 }
 
+
 __device__ Face::Face(const SpacePoint *vertices, int n_of_vertices) :
         vertices(malloc_and_copy(vertices, n_of_vertices)), n_of_vertices(n_of_vertices),
         normal(calculate_normal(vertices)), node(nullptr)
 {
 
 }
 
-__device__ Face::Face(const Face &other) :
-        vertices(malloc_and_copy(other.vertices, other.n_of_vertices)),
-        n_of_vertices(other.n_of_vertices),
-        normal(other.normal), node(other.node)
+__device__ Face &Face::operator=(const Face &other)
+{
+    if(this != &other)
+    {
+        vertices = malloc_and_copy(other.vertices, other.n_of_vertices);
```

Comment:
Please, note, that the actual type is `const SpacePoint *` (`*` is an important part). It is different from the `const SpacePoint *const` type (which would mean the field `vertices` can't be modified), but the actual time means that the **content** of the array can't be modified without replacing the whole array. If the field was const pointer, it would take away the opportunity to, for example, move a face into an existing one, but the pointer to const only prevents us from accidentally modifying the vertices array (vertices of an existing `Face` shouldn't change i. e. have a specific set of vertices, shouldn't they?). Because the field is private, it's already impossible to modify the array from outside of the `Face` class, but is still possible to modify it from inside, which shouldn't happen and is worth protecting.

The general rule is if you don't modify a variable, make it constant. This way you will get additional protection from your compiler